### PR TITLE
Documentation of lightning payments

### PR DESCRIPTION
**Pull request by akorosmezey**

### CODE REVIEW

**Comment by akorosmezey** [comment]
Open questions, not to forget:
1. Verify that the payment created in lock_spend is a return payment that is activated when something goes wrong, e.g., the secret is not revealed. 
Why is the the amount in the payment forwarded negative?

File: docs/design/lightning_payments.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
+# Lightning payments
+
+
+Send money to someone on the network that you are not sharing a state channel with.
+If you have a path of channels to the recipient you can use the nodes in between as 
+proxies in order to trustlessly send money to the recipient.
```
Comment (sennui) [inline docs/design/lightning_payments.md]: did you see implementation of that? Is there a table of channel pairs - equivalent of routing table, that would enable such a feature?

File: docs/design/lightning_payments.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+#### Make a locked payment
+    `channel_feeder:make_locked_payment/4` with input: pubkey of the proxy node,
+    amount + fee, the lightning code and an empty proof.
+1.  take the highest-nonced SPK of our node from channel data (#cd.me)
+2.  create new bet with
+    *   the code,
+    *   code as key
+    *   amount of payment and
+    *   an empty proof
+3.  apply the bet (i.e., add it to bets of the SPK)
```
Comment (sennui) [inline docs/design/lightning_payments.md]: we we call it a bet? is the code re-used with martket implementation?

File: docs/design/lightning_payments.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+5.  sign the new SPK
+
+#### Send it to server via talker (locked_payment message)
+    We are giving money conditionally, and asking us to forward it with a similar
+    conditions to the receiver. The message passes on parameters of the payment,
+    including the bet code and the code/secret pair encrypted with the pubkey of
+    the recipient.
+    The talker message processed on the proxy node:
+1.  `channel_feeder:lock_spend/7`
+    1.  First check that this channel is in the on-chain state with sufficient depth
```
Comment (sennui) [inline docs/design/lightning_payments.md]: what is depth? did you mean funds?

File: docs/design/lightning_payments.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        *   fail if amount <= 0
+        *   fail if fee <= lightning fee
+    3.  create signed return payment SPK (sanity check: test that same as the
+        original one without the signature)
+    4.  update channel data on the channel manager to
+        * me: received (unsigned) SPK
+        * them: received signed SPK
+        * add empty binary to ScriptSig of both sides
```
Comment (sennui) [inline docs/design/lightning_payments.md]: why we do that?

File: docs/design/lightning_payments.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    *   CID
+    *   accounts of the two endpoints
+    *   account 2
+    *   bets to be played
+    *   space and time gas (fees to be paid for storage and execution)
+    *   amount: money paid deposited into the channel by participants when the channel is
+        created and used during execution
+    *   nonce
+    *   entropy
+    *   delay: the maximum time to get the money out of the channel. Used in order to have
```
Comment (sennui) [inline docs/design/lightning_payments.md]: is it similar to timelock?

Comment (akorosmezey) [inline docs/design/lightning_payments.md]: The 'how_to_use_channels.md' says that, when my partner tries to close the channel and steal my money I have this much time to intervene. So this is the time one needs to wait until the channel can be closed. Thank you, I clarified this description.

File: docs/design/lightning_payments.md [hunk 6]
```diff
@@ -19,66 +12,81 @@ secret of the hashlock.
...
+    Call `api:lightning_spend/5` with inputs: proxy node IP, port, receiver's pubkey, amount, fee
+
 #### Create new lightning
-    `secrets:new_lightning/0`
+We first grab the [scriptpubkey (SPK) and scriptsig (SS) scripts](smart_contracts.md) from `secrets:new_lightning/0`:
 1.  generate random (32) bytes, this will be used as a secret
 2.  calculate the hash of it
 3.  generate a compiled chalang **code** referring to the hash of the random bytes
+    the returns repares a stack of 4 values: 
+        *   shared root: always []
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Formatting seems not optimal on GitHub.

File: docs/design/lightning_payments.md [hunk 7]
```diff
@@ -19,66 +12,81 @@ secret of the hashlock.
...
+    Create payment:
+    Call `api:lightning_spend/5` with inputs: proxy node IP, port, receiver's pubkey, amount, fee
+
 #### Create new lightning
-    `secrets:new_lightning/0`
+We first grab the [scriptpubkey (SPK) and scriptsig (SS) scripts](smart_contracts.md) from `secrets:new_lightning/0`:
 1.  generate random (32) bytes, this will be used as a secret
 2.  calculate the hash of it
 3.  generate a compiled chalang **code** referring to the hash of the random bytes
+    the returns repares a stack of 4 values:
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Please review wording.

File: docs/design/lightning_payments.md [hunk 8]
```diff
@@ -169,3 +254,6 @@ We assume a channel is already opened to the proxy node with sufficient funds fo
 
 ### gas
     Fee for using resources: space gas for storage and time-gas for execution.
+
+### Chalang
+    [Chalang](https://github.com/zack-bitcoin/chalang) a stack-based language implementing smart contracts.
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Please point to canonical repo at https://github.com/aeternity/chalang/ (reference rebar.config)

File: docs/design/lightning_payments.md [hunk 9]
```diff
@@ -19,66 +12,81 @@ secret of the hashlock.
...
 3.  generate a compiled chalang **code** referring to the hash of the random bytes
+    the returns repares a stack of 4 values: 
+        *   shared root: always []
+        *   amount 0 or 1 times channel granularity
+        *   nonce
+        *   delay
+    code:
+    *   if the stack is empty: [[], 0, 1, 50]
+    *   otherwise calculate hash of its top
+        *   if matches the hash wired-in: [[], granularity, 2, 0]
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: @akorosmezey Following our chat - well spotted that the [chalang code](https://github.com/aeternity/testnet/blob/32c888e8742f7ed75d15217460a772423dc3984a/apps/ae_core/src/consensus/secrets.erl#L46) appears to refer to the [constant channel granularity](https://github.com/aeternity/testnet/blob/32c888e8742f7ed75d15217460a772423dc3984a/apps/ae_core/src/consensus/constants.erl#L172-L173). A standalone refactoring PR - with description of context and rational behind - would be helpful for the code readability.

File: docs/design/lightning_payments.md [hunk 10]
```diff
@@ -19,66 +12,81 @@ secret of the hashlock.
...
 1.  generate random (32) bytes, this will be used as a secret
 2.  calculate the hash of it
 3.  generate a compiled chalang **code** referring to the hash of the random bytes
+    the returns repares a stack of 4 values: 
+        *   shared root: always []
+        *   amount 0 or 1 times channel granularity
+        *   nonce
+        *   delay
+    code:
+    *   if the stack is empty: [[], 0, 1, 50]
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: @akorosmezey Following our chat, I understand this is a *sanity check*, as the chalang code (as described in following bullet point) expects one input i.e. the alleged secret. You may consider adding a note on this line mentioning that this is meant to be a sanity check.

File: docs/design/lightning_payments.md (hunk 4 above)
Comment (akorosmezey) [inline docs/design/lightning_payments.md]: It is a placeholder, we are supposed to have equal number of bets and SS entries.

File: docs/design/lightning_payments.md (hunk 7 above)
Comment (akorosmezey) [inline docs/design/lightning_payments.md]: thanks, I fixing it

File: docs/design/lightning_payments.md (hunk 6 above)
Comment (akorosmezey) [inline docs/design/lightning_payments.md]: Fixing indentation

File: docs/design/lightning_payments.md (hunk 8 above)
Comment (akorosmezey) [inline docs/design/lightning_payments.md]: fixing, thanks

**Comment by akorosmezey** [comment]
I am trying to close this PR and add further improvements as separate smaller PRs as it grew too long.

File: docs/design/lightning_payments.md [hunk 11]
```diff
@@ -1,15 +1,244 @@
...
+
+Create payment by calling `api:lightning_spend/5` with inputs: proxy node IP, port, receiver's pubkey, amount and transaction fee
+
+#### Create new lightning
+
+We first grab the [scriptpubkey (SPK) and scriptsig (SS) scripts](smart_contracts.md) from `secrets:new_lightning/0`:
+
+1.  generate random (32) bytes, this will be used as a secret
+2.  calculate the hash of it
+3.  generate a compiled chalang **code** referring to the hash of the random bytes. It returns a stack of 4 values:
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Minor formatting issue.

File: docs/design/lightning_payments.md [hunk 12]
```diff
@@ -1,15 +1,244 @@
...
+Create payment by calling `api:lightning_spend/5` with inputs: proxy node IP, port, receiver's pubkey, amount and transaction fee
+
+#### Create new lightning
+
+We first grab the [scriptpubkey (SPK) and scriptsig (SS) scripts](smart_contracts.md) from `secrets:new_lightning/0`:
+
+1.  generate random (32) bytes, this will be used as a secret
+2.  calculate the hash of it
+3.  generate a compiled chalang **code** referring to the hash of the random bytes. It returns a stack of 4 values: 
+        *   shared root: always []
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Minor formatting issues in this list.

File: docs/design/lightning_payments.md [hunk 13]
```diff
@@ -1,15 +1,244 @@
...
+    *   fixed 1000 time-gas and
+    *   fixed 1000 space-gas
+4.  (sanity check: match trees in `tx_pool:data()`)
+5.  sign the new SPK and return it to `api:lightning_spend/7`.
+
+#### Send it to server via talker (locked_payment message)
+
+We are giving money conditionally, and asking us to forward it with similar conditions to the receiver. The message passes on parameters of the payment, including the bet code and the code/secret pair encrypted with the pubkey of the recipient.
+
+We tell the proxy node to make a locked payment, too, via `talker:talk`. On that node it goes to `ext_handler:doit({locked_payment` and then to `channel_feeder:handle_call({lock_spend`. The node cretaes two locked payments: one debiting the sender and one crediting the receiver:
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Minor typo "cretaes".

File: docs/design/lightning_payments.md [hunk 14]
```diff
@@ -1,15 +1,244 @@
...
+                        *   `spk:run/11` for the remainder
+                *   return SPK's original + collected amount, nonce, Shares, maximum of delays
+        *   Check nonces:
+            *   unless new one is lower, to run all bets with the new SS'es and return
+                the new bets, SS'es amount and nonce
+            *   if higher,
+                `spk:simplify_helper/2`
+                    try to unlock bets in in the SPK of the CD in my side using their SS
+                    `spk:bet_unlock/2`
+                    *   `spk:bet_unlock/8`
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Minor formatting issue on Github.

File: docs/design/lightning_payments.md [hunk 15]
```diff
@@ -1,15 +1,244 @@
...
+    *   update my side of channel data with the new SPK 
+        `channel_feeder:update_to_me/2`
+
+## Problems
+
+1.  bet evaluation: uses granularity from constants but the chalang code in secrets.erl has it wired-in.
+
+2.  chalang: duplication of opcodes, error-prone
+
+3.  almost all tests rely on badmatch/crash
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Which tests?

File: docs/design/lightning_payments.md [hunk 16]
```diff
@@ -1,15 +1,244 @@
...
+    A pubkey identifying a participant.
+
+### channel
+    Connection between two participants, created in order to perform some transctions.
+
+### CID
+    Channel ID. A network-wide unique number assigned to the channel.
+
+### bet
+    Contains:
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Formatting on github.

File: docs/design/lightning_payments.md [hunk 17]
```diff
@@ -1,15 +1,244 @@
...
+    *   code to execute
+    *   amount that goes to one of the participant depending on the outcome
+    *   prove
+    *   key: used to pattern-match that use the same contract. aiding channel updating.
+        The code itself is used as key for lightning payments.
+
+### SPK, script pubkey
+    [SPK, script pubkey](https://bitcoin.org/en/glossary/pubkey-script) is a Bitcoin term.
```
Comment (lucafavatella) [inline docs/design/lightning_payments.md]: Formatting on github.

Comment (lucafavatella) [inline docs/design/lightning_payments.md]: also in other items below.

File: docs/design/lightning_payments.md (hunk 9 above)
Comment (akorosmezey) [inline docs/design/lightning_payments.md]: @lucafavatella PR 333 created in order to replace the value 10000 with constants:channel_granularity(). Docs being updated, thanks.

File: docs/design/lightning_payments.md (hunk 10 above)
Comment (akorosmezey) [inline docs/design/lightning_payments.md]: @lucafavatella adding explanation

File: docs/design/lightning_payments.md (hunk 15 above)
Comment (akorosmezey) [inline docs/design/lightning_payments.md]: @lucafavatella I meant sanity checks, letting crash if a condition not satisfied, clarifying it, thanks.

File: docs/design/lightning_payments.md (hunk 17 above)
Comment (akorosmezey) [inline docs/design/lightning_payments.md]: @lucafavatella Hopefully fixed now

**Pull request merged**